pallets/order-book/src/types.rs

```diff
@@ -186,11 +224,169 @@ impl<AssetId: PartialEq> DealInfo<AssetId> {
+            for (account, amount) in from_whom.iter() {
+                Locker::lock_liquidity(
+                    self.dex_id,
+                    account,
+                    self.order_book_id,
+                    asset_id,
+                    *amount,
+                )?;
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn unlock<Unlocker>(&self) -> Result<(), DispatchError>
+    where
+        Unlocker: CurrencyUnlocker<AccountId, AssetId, DEXId, DispatchError>,
+    {
+        for (asset_id, to_whom) in self.to_unlock.iter() {
+            Unlocker::unlock_liquidity_batch(self.dex_id, self.order_book_id, asset_id, to_whom)?;
+        }
+
+        Ok(())
+    }
+
+    pub fn execute_all<Locker, Unlocker>(&self) -> Result<(), DispatchError>
+    where
+        Locker: CurrencyLocker<AccountId, AssetId, DEXId, DispatchError>,
+        Unlocker: CurrencyUnlocker<AccountId, AssetId, DEXId, DispatchError>,
+    {
+        self.lock::<Locker>()?;
+        self.unlock::<Unlocker>()?;
+        Ok(())
+    }
+}
+
+#[derive(Eq, PartialEq, Clone, RuntimeDebug)]
+pub struct MarketChange<AccountId, AssetId, DEXId, OrderId, LimitOrder, BlockNumber> {
+    // Info fields
+    /// The amount of the input asset for the exchange deal
+    pub deal_input: Option<OrderAmount>,
+
+    /// The amount of the output asset for the exchange deal
+    pub deal_output: Option<OrderAmount>,
+
+    /// The amount of the input asset that is placed into the market
+    pub market_input: Option<OrderAmount>,
+
+    /// The amount of the output asset that is placed into the market
+    pub market_output: Option<OrderAmount>,
+
+    // Fields to apply
+    pub to_add: BTreeMap<OrderId, LimitOrder>,
+    pub to_update: BTreeMap<OrderId, OrderVolume>,
     /// order id and number of block it is scheduled to expire at
-    pub to_delete: Vec<(OrderId, BlockNumber)>,
-    pub to_update: Vec<LimitOrder>,
-    pub makers_output: BTreeMap<AccountId, OrderVolume>,
+    pub to_delete: BTreeMap<OrderId, BlockNumber>,
+    pub payment: Payment<AssetId, AccountId, DEXId>,
+    pub ignore_unschedule_error: bool,
+}
+
+impl<AccountId, AssetId, DEXId, OrderId, LimitOrder, BlockNumber>
+    MarketChange<AccountId, AssetId, DEXId, OrderId, LimitOrder, BlockNumber>
+where
+    AssetId: Copy + PartialEq + Ord,
+    AccountId: Ord + Clone,
+    DEXId: Copy + PartialEq,
+    OrderId: Copy + Ord,
+{
+    pub fn new(dex_id: DEXId, order_book_id: OrderBookId<AssetId>) -> Self {
+        Self {
+            deal_input: None,
+            deal_output: None,
+            market_input: None,
+            market_output: None,
+            to_add: BTreeMap::new(),
+            to_update: BTreeMap::new(),
+            to_delete: BTreeMap::new(),
+            payment: Payment::new(dex_id, order_book_id),
+            ignore_unschedule_error: false,
+        }
+    }
+
+    pub fn merge(&mut self, mut other: Self) -> Result<(), ()> {
+        let join = |lhs: Option<OrderAmount>,
```

Comment:
also maybe just `fn`

Reply:
The logic is not complicated, the closure is suitable. Also if we move it to the separate function, it becomes possible to reuse it, but I want to avoid it because it is merge-specific behavior

Reply:
Closure adds more unnecessary syntax here and I meant to define `fn` in the scope of merge [like so](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=db46f931086de704a2e89e40383744f4)